pil_analyzer/src/evaluator.rs

```diff
@@ -1,104 +1,383 @@
-use std::collections::HashMap;
+use std::{collections::HashMap, fmt::Display, iter::repeat, rc::Rc};
 
 use ast::{
-    analyzed::{Expression, FunctionValueDefinition, Reference, Symbol, SymbolKind},
+    analyzed::{Expression, FunctionValueDefinition, Reference, Symbol},
     evaluate_binary_operation, evaluate_unary_operation,
     parsed::{FunctionCall, MatchArm, MatchPattern},
 };
+use itertools::Itertools;
 use number::FieldElement;
 
-/// Evaluates an expression to a single value.
+/// Evaluates an expression to a single number.
 pub fn evaluate_expression<T: FieldElement>(
-    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition<T>>)>,
     expression: &Expression<T>,
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition<T>>)>,
+) -> Result<T, String> {
+    internal::evaluate_to_number::<T, FixedColumnRef<'_, T>>(
+        expression,
+        &[],
+        &Definitions(definitions),
+    )
+}
+
+pub fn evaluate<'a, T: FieldElement, C: Custom<Value<'a, T, C>>>(
+    expr: &'a Expression<T>,
+    symbols: &impl SymbolLookup<'a, T, C>,
+) -> Result<Value<'a, T, C>, String> {
+    internal::evaluate(expr, &[], symbols)
+}
+
+pub fn evaluate_function_call<'a, T: FieldElement, C: Custom<Value<'a, T, C>>>(
+    expr: &'a Expression<T>,
+    arguments: Vec<T>,
+    symbols: &impl SymbolLookup<'a, T, C>,
+) -> Result<Value<'a, T, C>, String> {
+    internal::evaluate(
+        expr,
+        &arguments
+            .into_iter()
+            .map(|x| Rc::new(Value::Number(x)))
+            .collect::<Vec<_>>(),
+        symbols,
+    )
+}
+
+pub fn evaluate_to_number<'a, T: FieldElement, C: Custom<Value<'a, T, C>>>(
+    expr: &'a Expression<T>,
+    symbols: &impl SymbolLookup<'a, T, C>,
+) -> Result<T, String> {
+    internal::evaluate_to_number(expr, &[], symbols)
+}
+
+pub fn evaluate_function_call_to_number<'a, T: FieldElement, C: Custom<Value<'a, T, C>>>(
+    expr: &'a Expression<T>,
+    arguments: Vec<T>,
+    symbols: &impl SymbolLookup<'a, T, C>,
 ) -> Result<T, String> {
-    Evaluator {
-        definitions,
-        function_cache: &Default::default(),
-        variables: &[],
+    internal::evaluate_to_number(
+        expr,
+        &arguments
+            .into_iter()
+            .map(|x| Rc::new(Value::Number(x)))
+            .collect::<Vec<_>>(),
+        symbols,
+    )
+}
+
+#[derive(Clone, PartialEq)]
+pub enum Value<'a, T, C> {
+    Number(T),
+    String(String),
+    Tuple(Vec<Self>),
+    Array(Vec<Self>),
+    Closure(Closure<'a, T, C>),
+    Custom(C),
+}
+
+// TODO PartialEq is only needed because don't have proper patterns in match expressions.
+pub trait Custom<V>: Display + Clone + PartialEq {
+    fn eval_function_application(self, arguments: &[Rc<V>]) -> Result<V, String>;
```

Comment:
Maybe it would be better to move this into the Symbols trait and keep the Custom trait fuction-less (i.e. reduce it to the template parameter in Value).